File: bin/dcgo.r
library(stringr)
library(dcGOR)
Pfam <- dcRDataLoader('Pfam')
background_ids = c()
domtblout_file = readLines("../results/dcgo/seqdb_Pfam-A.domtblout")
for (line in domtblout_file)
{
  match = str_match(line, "PF[:digit:]+")
  if(!is.na(match))
  {
    background_ids = c(background_ids, match)
  }
}

pfam_ids = c()
domtblout_file = readLines("../results/dcgo/beneficial-loss_pfam-A.domtblout")
for (line in domtblout_file)
{
  match = str_match(line, "PF[:digit:]+")
  if(!is.na(match))
  {
    pfam_ids = c(pfam_ids, match)
  }
}
benloss <- dcEnrichment(pfam_ids, domain="Pfam", ontology="GOBP", background=background_ids)
# pdf("../result/core-essential.pdf")
# visEnrichment(benloss, num_top_nodes = 20)
# dev.off()
a <- view(benloss, top_num=NA, sortBy="adjp", details=TRUE)
b = cbind.data.frame(a$term_id, a$term_name, a$adjp)
colnames(b) = c("Go", "name", "adjusted_P-value")
c= b[b$`adjusted_P-value`<0.01,]
write.table(c, file="../results/dcgo/beneficial-loss.txt", row.names=FALSE, quote = FALSE)

pfam_ids = c()
domtblout_file = readLines("../results/dcgo/non-essential_pfam-A.domtblout")
for (line in domtblout_file)
{
  match = str_match(line, "PF[:digit:]+")
  if(!is.na(match))
  {
    pfam_ids = c(pfam_ids, match)
  }
}
noness <- dcEnrichment(pfam_ids, domain="Pfam", ontology="GOBP", background=background_ids)
# pdf("../result/core-essential.pdf")
# visEnrichment(noness, num_top_nodes = 20)
# dev.off()
a <- view(noness, top_num=NA, sortBy="adjp", details=TRUE)
b = cbind.data.frame(a$term_id, a$term_name, a$adjp)
colnames(b) = c("Go", "name", "adjusted_P-value")
c= b[b$`adjusted_P-value`<0.01,]
write.table(c, file="../results/dcgo/non-essential.txt", row.names=FALSE, quote = FALSE)

pfam_ids = c()
domtblout_file = readLines("../results/dcgo/essential_pfam-A.domtblout")
for (line in domtblout_file)
{
  match = str_match(line, "PF[:digit:]+")
  if(!is.na(match))
  {
    pfam_ids = c(pfam_ids, match)
  }
}
ess <- dcEnrichment(pfam_ids, domain="Pfam", ontology="GOBP", background=background_ids)
# pdf("../result/core-essential.pdf")
# visEnrichment(noness, num_top_nodes = 20)
# dev.off()
a <- view(ess, top_num=NA, sortBy="adjp", details=TRUE)
b = cbind.data.frame(a$term_id, a$term_name, a$adjp)
colnames(b) = c("Go", "name", "adjusted_P-value")
c= b[b$`adjusted_P-value`<0.01,]
write.table(c, file="../results/dcgo/essential.txt", row.names=FALSE, quote = FALSE)

File: bin/dcgo.py
from os import listdir

def read_essentiality(filespath):
    list_of_files = listdir(filespath)
    ess_dict = {}
    for filename in list_of_files:
        with open(filespath + filename) as fromfile:
            for line in fromfile:
                cells = line.split()
                if cells[2] in ['essential', 'non-essential', 'beneficial-loss']:
                    ess_dict[cells[0]] = cells[2]
    return ess_dict

essentiality = '../results/biases/dbscan/'
ess_dict = read_essentiality(essentiality)
hmmscan = '../results/dcgo/seqdb_Pfam-A.domtblout'
essen = '../results/dcgo/essential_pfam-A.domtblout'
noness = '../results/dcgo/non-essential_pfam-A.domtblout'
benloss = '../results/dcgo/beneficial-loss_pfam-A.domtblout'

with open(hmmscan, 'r') as fromfile:
    with open(essen, 'w') as essenfile:
        with open(noness, 'w') as nonessfile:
            with open(benloss, 'w') as benlossfile:
                for line in fromfile:
                    if not line.startswith('#'):
                        cells = line.split()
                        if cells[3] in ess_dict:
                            if ess_dict[cells[3]] == 'essential':
                                essenfile.write(line)
                            elif ess_dict[cells[3]] == 'non-essential':
                                nonessfile.write(line)
                            else:
                                benlossfile.write(line)


